File: Graph/332. Reconstruct Itinerary.py
# -*- coding: utf-8 -*-
"""
Created on Sun Nov 11 20:07:42 2018

@author: TZLMYQ
"""
import collections
class Solution1(object):
    def findItinerary(self, tickets):
        """
        :type tickets: List[List[str]]
        :rtype: List[str]
        """
        self.d = collections.defaultdict(list)
        res = []
        for ticket in tickets:
            self.d[ticket[0]].append(ticket[1])
        for key in self.d:
            self.d[key].sort(reverse=True)
        stack = ["JFK"]
        
        while stack:
            city = stack[-1]
            if len(self.d[city]) == 0:
                res.append(city)
                stack.pop()
            else:
                stack.append(self.d[city].pop())
        return res[::-1]
    
class Solution2:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        self.d = collections.defaultdict(list)
        for ticket in tickets:
            self.d[ticket[0]].append(ticket[1])
        for key in self.d:
            self.d[key].sort(reverse=True)
        self.res = []
        self.dfs("JFK", tickets)
        return self.res[::-1]
    
    def dfs(self, start, tickets):
        if start in self.d and len(self.d[start]) > 0:
            while len(self.d[start]) > 0:
                next_stop = self.d[start].pop()
                self.dfs(next_stop, tickets)
        self.res.append(start)
        
from collections import deque
class Solution3:
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        self.d = collections.defaultdict(list)
        for flight in tickets:
            self.d[flight[0]].append(flight[1])
        self.route = ["JFK"]
        length = len(tickets)
        self.dfs("JFK", length)
        return self.route
    
    def dfs(self, start, k):
        if len(self.route) == k + 1:
            return self.route
        cities = sorted(self.d[start])
        for city in cities:
            self.d[start].remove(city)
            self.route.append(city)
            valid = self.dfs(city, k)
            if valid:
                return valid
            self.route.pop()
            self.d[start].append(city)
 

    
solution = Solution()
#tickets = [["MUC","LHR"],["JFK","MUC"],["SFO","SJC"],["LHR","SFO"]]
tickets = [["JFK","SFO"],["JFK","ATL"],["SFO","ATL"],["ATL","JFK"],["ATL","SFO"]]
print(solution.findItinerary(tickets))

File: DFS/332. Reconstruct Itinerary.cpp
// 这是一道超级好的题，
// 1.DFS的思想非常的秒。两个stack左手倒腾到右手。
// 2.multiset的用法非常好。尤其是在第27行，需要注意，m[airport].erase(m[airport].begin()) 和m[airport].erase(*m[airport].begin())都是有效的，但是意义不同。
// 3. multiset 本来就是一个set里面含有多个重复的元素，既可以一次删除所有，还可以一次删除一个。
// 4. 那么，为什么这个时候我们不能够用vector<string>来代替这里的multiset呢，在下面再加上一行sort(v.begin(), v.end())，不行嘛？行，但是千万要注意加上“&”。不然排序的是copy而不是reference。
class Solution {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        unordered_map<string, multiset<string>> m;
        for(auto ticket:tickets)
            m[ticket[0]].insert(ticket[1]);
        vector<string> res;
        vector<string> stack;
        
        stack.push_back("JFK");
        while(!stack.empty())
        {
            string airport = stack.back();
            if(m[airport].empty())
            {
                res.push_back(airport);
                stack.pop_back();                
            }
            else
            {
                stack.push_back(*m[airport].begin());
                m[airport].erase(m[airport].begin());
            }
        }
        return vector<string>(res.rbegin(), res.rend());
    }
};

class Solution1 {
public:
    vector<string> findItinerary(vector<vector<string>>& tickets) {
        unordered_map<string, vector<string>> m;
        for(auto ticket:tickets)
            m[ticket[0]].push_back(ticket[1]);
        for(auto& it:m)
            sort(it.second.begin(), it.second.end());
        vector<string> res;
        vector<string> stack;
        
        stack.push_back("JFK");
        while(!stack.empty())
        {
            string airport = stack.back();
            if(m[airport].empty())
            {
                res.push_back(airport);
                stack.pop_back();                
            }
            else
            {
                stack.push_back(*m[airport].begin());
                m[airport].erase(m[airport].begin());
            }
        }
        return vector<string>(res.rbegin(), res.rend());
    }
};
